### PR TITLE
[IOTDB-814] Cache and redirect distributed IoTDB's leader for write requests in IoTDB's client 

### PR DESCRIPTION
**Pull request by LebronAl**

### CODE REVIEW

**Comment by LebronAl** [comment]
# Benchmark Parameter
CLIENT_NUMBER=1
GROUP_NUMBER=1
DEVICE_NUMBER=20
SENSOR_NUMBER=20
BATCH_SIZE=50
LOOP=1000
DATA_TYPE=DOUBLE
INSERT_MODE=session
# Cluster config
default_replica_num=1
enablePartition=false
two nodes's cluster(34,35)
# Cluster_new's latest Version
## forward(benchmark writes data to 34, sg belongs to 35,internal forward)
image
## non-forward(benchmark writes data to 35, sg belongs to 35,no need to forward)
image
# Cluster_new_cache_leader's latest Version
## Close cache leader
###  forward(benchmark writes data to 34, sg belongs to 35,internal forward)
image
### non-forward(benchmark writes data to 35, sg belongs to 35,no need to forward)
image
## Open cache leader
### little forward(benchmark writes data to 34, sg belongs to 35,internal forward first write for each deviceId + cache leader)
image
### non-fowward(benchmark writes data to 34, sg belongs to 35,no need to forward)
image

**Comment by LebronAl** [comment]
@jt2594838 @neuyilan @kr11 @qiaojialin PTAL

File: session/src/main/java/org/apache/iotdb/session/SessionConnection.java [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+  public SessionConnection(Session session, EndPoint endPoint) throws IoTDBConnectionException {
+    this.session = session;
+    init(endPoint);
+  }
+
+  private void init(EndPoint endPoint) throws IoTDBConnectionException {
+    transport = new TFastFramedTransport(
+        new TSocket(endPoint.getIp(), endPoint.getPort(), session.connectionTimeoutInMs));
+
+    if (!transport.isOpen()) {
```
Comment (jt2594838) [inline session/src/main/java/org/apache/iotdb/session/SessionConnection.java:77]: I think this condition is redundant.

File: session/src/main/java/org/apache/iotdb/session/SessionConnection.java [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+    } catch (TException e) {
+      throw new IoTDBConnectionException(e);
+    }
+  }
+
+  protected boolean checkTimeseriesExists(String path)
+      throws IoTDBConnectionException, StatementExecutionException {
+    SessionDataSet dataSet = executeQueryStatement(String.format("SHOW TIMESERIES %s", path));
+    boolean result = dataSet.hasNext();
+    dataSet.closeOperationHandle();
```
Comment (jt2594838) [inline session/src/main/java/org/apache/iotdb/session/SessionConnection.java:211]: Close operation should be put into a finally block.

**Review by jt2594838 — CHANGES_REQUESTED** [review]

File: session/src/main/java/org/apache/iotdb/session/SessionConnection.java (hunk 2 above)
Comment (LebronAl) [reply session/src/main/java/org/apache/iotdb/session/SessionConnection.java:211]: OK~

File: session/src/main/java/org/apache/iotdb/session/SessionConnection.java (hunk 1 above)
Comment (LebronAl) [reply session/src/main/java/org/apache/iotdb/session/SessionConnection.java:77]: I copy this from old session's implemention.I will delete it anyway.

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 3]
```diff
@@ -691,42 +857,38 @@ public void testInsertRecord(String deviceId, long time, List<String> measuremen
    */
   public void deleteTimeseries(String path)
       throws IoTDBConnectionException, StatementExecutionException {
-    List<String> paths = new ArrayList<>();
-    paths.add(path);
-    deleteTimeseries(paths);
+    defaultSessionConnection.deleteTimeseries(new ArrayList<String>() {{
+      add(path);
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply session/src/main/java/org/apache/iotdb/session/Session.java:861]: OK~

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 4]
```diff
@@ -121,209 +105,326 @@ private synchronized void open(boolean enableRPCCompression, int connectionTimeo
...
+   * @see Session#insertTablet(Tablet)
    */
-  public void insertTablets(Map<String, Tablet> tablets)
-      throws IoTDBConnectionException, StatementExecutionException {
-    insertTablets(tablets, false);
+  public void insertRecord(String deviceId, long time, List<String> measurements,
+      List<TSDataType> types,
+      Object... values) throws IoTDBConnectionException, StatementExecutionException {
+    TSInsertRecordReq request = genTSInsertRecordReq(deviceId, time, measurements, types,
+        new ArrayList<>(Arrays.asList(values)));
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply session/src/main/java/org/apache/iotdb/session/Session.java:330]: ok~

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 5]
```diff
@@ -121,209 +105,326 @@ private synchronized void open(boolean enableRPCCompression, int connectionTimeo
...
+      }
+    }
+  }
 
-      RpcUtils.verifySuccess(openResp.getStatus());
+  public void deleteStorageGroup(String storageGroup)
+      throws IoTDBConnectionException, StatementExecutionException {
+    try {
+      metaSessionConnection.deleteStorageGroups(new ArrayList<String>() {{
+        add(storageGroup);
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply session/src/main/java/org/apache/iotdb/session/Session.java:166]: ok~

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 6]
```diff
@@ -121,209 +105,326 @@ private synchronized void open(boolean enableRPCCompression, int connectionTimeo
...
+  public void setStorageGroup(String storageGroup)
+      throws IoTDBConnectionException, StatementExecutionException {
     try {
-      TSOpenSessionResp openResp = client.openSession(openReq);
+      metaSessionConnection.setStorageGroup(storageGroup);
+    } catch (RedirectException e) {
+      if (Config.DEFAULT_CACHE_LEADER_MODE) {
+        logger.debug("storageGroup[{}]:{}", storageGroup, e.getMessage());
+        metaSessionConnection = new SessionConnection(this, e.getEndPoint());
+        endPointToSessionConnection.putIfAbsent(e.getEndPoint(), metaSessionConnection);
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply session/src/main/java/org/apache/iotdb/session/Session.java:157]: OK, I intend to update `endPointToSessionConnection ` first and then update `metaSessionConnection` from `endPointToSessionConnection `

File: cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java [hunk 7]
```diff
@@ -1115,7 +1118,11 @@ TSStatus executeNonQueryPlan(PhysicalPlan plan) {
         return status;
       }
     } else if (leader != null) {
-      return forwardPlan(plan, leader, getHeader());
+      TSStatus result = forwardPlan(plan, leader, getHeader());
+      if (!StatusUtils.NO_LEADER.equals(result)) {
+        result.setRedirectNode(new EndPoint(leader.getIp(), leader.getClientPort()));
+        return result;
+      }
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java:1125]: Actually I designed this on purpose. If the first call to the `forwardPlan` function returns StatusUtils.NO_LEADER. We should block this request until the new leader is elected. So the following code should be called. This is why I put this return statement in the scope of `if`. Of course,  If the second call to the `forwardPlan` function returns StatusUtils.NO_LEADER too, then maybe just return will be better~

File: cluster/src/main/java/org/apache/iotdb/cluster/server/member/MetaGroupMember.java [hunk 8]
```diff
@@ -1486,7 +1485,11 @@ private TSStatus processNonPartitionedMetaPlan(PhysicalPlan plan) {
         return status;
       }
     } else if (leader != null) {
-      return forwardPlan(plan, leader, null);
+      TSStatus result = forwardPlan(plan, leader, null);
+      if (!StatusUtils.NO_LEADER.equals(result)) {
+        result.setRedirectNode(new EndPoint(leader.getIp(), leader.getClientPort()));
+        return result;
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/MetaGroupMember.java:1491]: same as above~Anything can be discussed here~

File: cluster/src/main/java/org/apache/iotdb/cluster/server/member/MetaGroupMember.java [hunk 9]
```diff
@@ -1688,13 +1693,20 @@ private TSStatus forwardInsertTabletPlan(Map<PhysicalPlan, PartitionGroup> planG
...
       logger.debug("{}: from {},{},{}", name, entry.getKey(), entry.getValue(), tmpStatus);
       noFailure =
           (tmpStatus.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) && noFailure;
       isBatchFailure = (tmpStatus.getCode() == TSStatusCode.MULTIPLE_ERROR.getStatusCode())
           || isBatchFailure;
+      if (tmpStatus.isSetRedirectNode()) {
+        endPoint = tmpStatus.getRedirectNode();
+      } else {
+        allRedirect = false;
+      }
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/MetaGroupMember.java:1709]: Actually I designed this on purpose.If the last element of the tmpStatus do not set the RedirectNode, which means the coordinator can process at lease one plan locally, so the final allRedirect value is false, the server will not recomment client a new Node to redirect.
BTW, maybe in our user's most scenarios, forwarding the plan to second Node is better, I will fix this.

File: cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (neuyilan) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java:1125]: I got it, better add more comments here~

[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java:1125]: copy that!

**Comment by LebronAl** [comment]
# Benchmark Parameter
CLIENT_NUMBER=1
GROUP_NUMBER=1
DEVICE_NUMBER=20
SENSOR_NUMBER=20
BATCH_SIZE=50
LOOP=10000
DATA_TYPE=DOUBLE
INSERT_MODE=session
# Cluster config
default_replica_num=1
enablePartition=false
is_enable_raft_log_persistence=true
is_use_async_applier=true
two nodes's cluster(34,35),benchmark(33)
# Cluster_new 
## Version
b42b0f194 massive refactor: RaftMembers now care about less
## forward(benchmark writes data to 34, sg belongs to 35,internal forward)
image
image
image
image
## non-forward(benchmark writes data to 35, sg belongs to 35,no need to forward)
image
image
image
image
# Cluster_new_cache_leader 
## Version
229779f adjust close cache leader logic
## Close cache leader
###  forward(benchmark writes data to 34, sg belongs to 35,internal forward)
image
image
image
image
### non-forward(benchmark writes data to 35, sg belongs to 35,no need to forward)
image
image
image
image
## Open cache leader
### little forward(benchmark writes data to 34, sg belongs to 35,internal forward first write for each deviceId + cache leader)
image
image
image
image
### non-forward(benchmark writes data to 35, sg belongs to 35,no need to forward)
image
image
image
image

**Comment by LebronAl** [comment]
# Benchmark Parameter
CLIENT_NUMBER=20
GROUP_NUMBER=20
DEVICE_NUMBER=200
SENSOR_NUMBER=500
BATCH_SIZE=50
LOOP=200
DATA_TYPE=DOUBLE
INSERT_MODE=session

# Cluster config
default_replica_num=2
enablePartition=false
is_enable_raft_log_persistence=false
is_use_async_applier=true
three nodes's cluster(33,34,35),benchmark(32)

# Cluster_new
## Version
ef2c8a2a4  fix method name
## Result
image
image
image
image
# Cluster_new_cache_leader
## Version
b2e92ed68 Merge branch 'cluster_new' of github.com:Apache/incubator-iotdb into cluster_new_cache_leader
## Result
image
image
image
image

**Comment by LebronAl** [comment]
# Benchmark Parameter
IS_CLIENT_BIND=False
CLIENT_NUMBER=20
GROUP_NUMBER=20
DEVICE_NUMBER=200
SENSOR_NUMBER=500
BATCH_SIZE=50
LOOP=200
DATA_TYPE=DOUBLE
INSERT_MODE=session

# Cluster config
default_replica_num=2
enablePartition=false
is_enable_raft_log_persistence=false
is_use_async_applier=true
three nodes's cluster(33,34,35),benchmark(32)

# Cluster_new
## Version
ef2c8a2 fix method name
## Result
image
image
image
image

# Cluster_new_cache_leader
## Version
b2e92ed Merge branch 'cluster_new' of github.com:Apache/incubator-iotdb into cluster_new_cache_leader
## Result
image
image
image
image

**Comment by LebronAl** [comment]
# Benchmark Parameter
IS_CLIENT_BIND=False
CLIENT_NUMBER=20
GROUP_NUMBER=20
DEVICE_NUMBER=200
SENSOR_NUMBER=500
BATCH_SIZE=50
LOOP=200
DATA_TYPE=DOUBLE
INSERT_MODE=session

# Cluster config
default_replica_num=2
enablePartition=false
is_enable_raft_log_persistence=false
is_use_async_applier=true
three nodes's cluster(33,34,35),benchmark(32) write to 33

# Cluster_new
## Version
4c3395add refactor timer
## Result
image
image
image
image
image
image
# Cluster_new_cache_leader
## Version
7d12d5ba2 Merge branch 'cluster_new' of github.com:Apache/incubator-iotdb into cluster_new_cache_leader
## Result
image
image
image
image
image
image

**Review by jt2594838 — APPROVED** [review]

**Pull request merged**